engine/graphics/font.cpp

```diff
@@ -0,0 +1,171 @@
+    fontCharacter.mYOffset = spaceAboveCharPixels;
+
+    if (character != '\n')
+    {
+        // Since this uses a single character for each texture, stride is just the pixel width of the character
+        const int kStride = charWidth;
+        stbtt_MakeCodepointBitmap(&mFontInfo, alphaTexture.data(),
+                                    charWidth, charHeight, kStride,
+                                    mPixelScale, mPixelScale,
+                                    character);
+    }
+
+    // Bitmap returned is alpha only, and stored bottom left pixel first
+    // Convert to RGBA and flip it for OpenGL
+    const auto rgbTexture = image::ConvertAlphaImageToRGBA(alphaTexture);
+    fontCharacter.mBitmap = image::FlipImageVertically(rgbTexture, charWidth, charHeight, 4);
+    fontCharacter.mBitmapWidth  = charWidth;
+    fontCharacter.mBitmapHeight = charHeight;
+
+    mCharacters[character] = fontCharacter;
+    return fontCharacter;
+}
+
+FontString Font::GetCharacters(const std::string& text)
+{
+    std::vector<FontCharacter> fontCharacters {};
+
+    float totalTextWidth = 0.0f;
+    float totalTextHeight = 0.0f;
+    float currentLineHeight = 0.0f;
+    float currentX = 0.0f;
+    float currentY = 0.0f;
+
+    stbtt_InitFont(&mFontInfo, mFontData.data(), 0);
+
+    for (int i = 0; i < text.length(); i++)
+    {
+        const char character = text[i];
+        FontCharacter fontCharacter = GetCharacter(character);
+
+        if (character == '\n')
+        {
+            currentX = 0;
+            currentY += (mAscent - mDescent + mLineGap);
+        }
+        else
+        {
+            currentLineHeight = std::fabs(fontCharacter.mYOffset + fontCharacter.mBitmapHeight);
+
+            // Left side bearing is the amount of space between the previous character and this one
+            // The first character does not need to take this into account
+            if (i > 0)
+            {
+                currentX += fontCharacter.mLeftSideBearing;
+            }
+
+            fontCharacter.mYOffset += currentY;
+            // Offset the character to the center of the quad
+            fontCharacter.mXOffset = currentX + fontCharacter.mBitmapWidth / 2.0f;
+            fontCharacter.mYOffset += fontCharacter.mBitmapHeight / 2.0f;
+
+            currentX += fontCharacter.mGlyphWidth;
+        }
+
+        totalTextWidth = std::max(totalTextWidth, currentX);
+        totalTextHeight = std::max(totalTextHeight, currentY + currentLineHeight);
+
+        if (i < text.length() - 1)
+        {
+            // The kerning is the amount of space between a specific character pair
+            // This can only be calculated if the next character is known
+            const int kern = stbtt_GetCodepointKernAdvance(&mFontInfo, character, text[i + 1]);
+            currentX += std::fabs(kern * mPixelScale);
+        }
+
+        fontCharacters.push_back(fontCharacter);
+    }
+
+    return { fontCharacters, totalTextWidth, totalTextHeight };
```

Comment:
Designated initializers are going to make this so much nicer in the future:

https://en.cppreference.com/w/cpp/language/aggregate_initialization

```
return { .chars = fontCharacters, .width = totalTextWidth, .height = totalTextHeight };
```

so much more explicit and less chance for bugs. C++20!

Reply:
Ahh very cool! I had no idea.